homeassistant/components/tuya/__init__.py

```diff
@@ -0,0 +1,296 @@
+
+    for platform in TUYA_SUPPORT_HA_TYPE:
+        _LOGGER.debug("tuya async platform-->%s", platform)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+        hass.data[DOMAIN][TUYA_SETUP_PLATFORM].add(platform)
+
+    return True
+
+
+async def cleanup_device_registry(hass: HomeAssistant):
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    __device_registry = device_registry.async_get(hass)
+    device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+
+    for dev_id, device_entity in list(__device_registry.devices.items()):
+        for item in device_entity.identifiers:
+            if DOMAIN == item[0] and item[1] not in device_manager.device_map:
+                __device_registry.async_remove_device(dev_id)
+                break
+
+
+def remove_hass_device(hass: HomeAssistant, device_id: str):
+    """Remove device from hass cache."""
+    __device_registry = device_registry.async_get(hass)
+    for entity in list(__device_registry.devices.values()):
+        print(f"inde->>{entity.identifiers}")
+        if device_id in list(entity.identifiers)[0]:
+            __device_registry.async_remove_device(entity.id)
+
+
+async def async_setup(hass, config):
+    """Set up the Tuya integration."""
+    tuya_logger.setLevel(_LOGGER.level)
+    conf = config.get(DOMAIN)
+
+    _LOGGER.debug("Tuya async setup conf %s", conf)
+    if conf is not None:
+
+        async def flow_init() -> Any:
+            try:
+                result = await hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": SOURCE_IMPORT}, data=conf
+                )
+            except UnknownFlow as flow:
+                _LOGGER.error(flow.args)
+            except UnknownStep as step:
+                _LOGGER.error(step.args)
+            except ValueError as err:
+                _LOGGER.error(err.args)
+            _LOGGER.debug("Tuya async setup flow_init")
+            return result
+
+        hass.async_create_task(flow_init())
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unloading the Tuya platforms."""
+    _LOGGER.debug("integration unload")
+    unload = await hass.config_entries.async_unload_platforms(
+        entry, hass.data[DOMAIN]["setup_platform"]
+    )
+    if unload:
+        __device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+        __device_manager.mq.stop()
+        __device_manager.remove_device_listener(hass.data[DOMAIN][TUYA_MQTT_LISTENER])
+
+        hass.data.pop(DOMAIN)
+
+    return unload
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Async setup hass config entry."""
+    _LOGGER.debug("tuya.__init__.async_setup_entry-->%s", entry.data)
+
+    hass.data[DOMAIN] = {TUYA_HA_TUYA_MAP: {}, TUYA_HA_DEVICES: []}
+    hass.data[DOMAIN][TUYA_SETUP_PLATFORM] = set()
```

Comment:
Please index this by config entry:
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
      TUYA_HA_TUYA_MAP: {},
      TUYA_HA_DEVICES: [],
      TUYA_SETUP_PLATFORM: set()
    }
```